onadata/libs/authentication.py

```diff
@@ -185,3 +194,67 @@ def authenticate(self, request):
             return None
 
         return self.authenticate_credentials(key)
+
+
+def check_lockout(request):
+    try:
+        if isinstance(request.META['HTTP_AUTHORIZATION'], bytes):
+            username = \
+                request.META['HTTP_AUTHORIZATION'].decode('utf-8').split('"')[
+                    1]
+        else:
+            username = request.META['HTTP_AUTHORIZATION'].split('"')[1]
+    except (TypeError, AttributeError, IndexError):
+        return
+    else:
+        lockout = cache.get('{}{}'.format(LOCKOUT_USER, username))
+        if lockout:
+            time_locked_out = \
+                datetime.now() - datetime.strptime(lockout,
+                                                   '%Y-%m-%dT%H:%M:%S')
+            remaining_time = round(
+                (settings.LOCKOUT_TIME - time_locked_out.seconds) / 60)
+            raise AuthenticationFailed(
+                _('Locked out. Too many password attempts. '
+                  'Try again in {} minutes'.format(remaining_time)))
+        return username
+
+
+def login_attempts(request):
+    """Track number of login attempts made by user within a specified amount
+     of time"""
+    username = check_lockout(request)
+    attempts = cache.get('{}{}'.format(LOGIN_ATTEMPTS, username))
+
+    if attempts:
+        cache.incr('{}{}'.format(LOGIN_ATTEMPTS, username))
+        attempts = cache.get('{}{}'.format(LOGIN_ATTEMPTS, username))
+        if attempts >= settings.MAX_LOGIN_ATTEMPTS:
+            send_lockout_email(username)
+            return cache.set(
+                '{}{}'.format(LOCKOUT_USER, username),
+                datetime.now().strftime('%Y-%m-%dT%H:%M:%S'),
+                settings.LOCKOUT_TIME)
+        return
+
+    return cache.set('{}{}'.format(LOGIN_ATTEMPTS, username), 1)
+
+
+def send_lockout_email(username):
+    try:
+        user = User.objects.get(username=username)
+    except User.DoesNotExist:
+        pass
+    else:
+        email_data = get_account_lockout_email_data(username)
+        end_email_data = get_account_lockout_email_data(
+            username, end=True)
+
+        send_account_lockout_email.apply_async(
+            args=[user.email, email_data.get('subject'),
+                  email_data.get('message_txt')])
+        # send end of lockout email 1 minute after lockout time
+        send_account_lockout_email.apply_async(
+            args=[user.email, end_email_data.get('subject'),
+                  end_email_data.get('message_txt')],
+            countdown=settings.LOCKOUT_TIME + 60)
```

Comment:
Access to a setting that does not have a default value in our default settings. We could perhaps have a constant on this module that would use `getattr()` instead?
```python

LOCKOUT_DURATION = getattr(settings, 'LOCKOUT_TIME', 3600)
```

It could also apply to `MAX_LOGIN_ATTEMPTS`.